### PR TITLE
Bugfix FXIOS-10482 Improve memory usage during tab tray open/close

### PR DESCRIPTION
**Pull request by mattreaganmozilla**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-10482)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/22952)

## :bulb: Description

Bandaid for a memory leak that occurs after opening/closing the tab tray. A general memory comparison graph (before/after) is shown below. 

The underlying bug is still a problem unfortunately, and I believe we are still leaking a few of the container views higher up in the view hierarchy. Additionally, I'm seeing some other related scenarios that need further investigation and also might be leaking other UI.

This PR, however, does free up memory use considerably in this basic scenario, so I'm opening this for review now so we can have this available in the meantime until there is a better longterm solution in place.

**Note**: this appears to be a longstanding issue with the tab tray that predates any of the recent UI updates.

---

### Memory Use Comparison

For this test the app was launched and then the tab tray was opened and closed repeatedly.

Allocations on device (iPhone 12 mini)

**Before**
Screenshot 2024-11-08 at 1 16 16 PM

**After**
Screenshot 2024-11-08 at 1 14 44 PM

**Before**
before

**After**
after

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

**Review by PARAIPAN9 — APPROVED** [review]
Definitely an improvement, @mattreaganmozilla. I see now an approximate 1MB increase in memory compared to 7MB previously, so an approximate of 7 times improvement. Really nice job.

**Comment by mattreaganmozilla** [comment]
> I see now an approximate 1MB increase in memory compared to 7MB previously

@PARAIPAN9 What steps were you using to test? I'm still investigating so I'll take a closer look; 1MB is still a big memory leak and I'm actually not seeing that with the fix in place.

**Review by dataports — APPROVED** [review]

File: firefox-ios/Client/Coordinators/TabTray/TabTrayCoordinator.swift [hunk 1]
```diff
@@ -33,6 +33,11 @@ class TabTrayCoordinator: BaseCoordinator,
         initializeTabTrayViewController(selectedTab: tabTraySection)
     }
 
+    func dismissChildTabTrayPanels() {
+        let childVCs: [UIViewController] =  tabTrayViewController.childPanelControllers.flatMap { $0.viewControllers }
```
Comment (dataports) [inline firefox-ios/Client/Coordinators/TabTray/TabTrayCoordinator.swift:37]: little nit, looks like there's an extra space on the rhs of the equals here

Comment (mattreaganmozilla) [reply firefox-ios/Client/Coordinators/TabTray/TabTrayCoordinator.swift]: Thank you, fixed

**Comment by PARAIPAN9** [comment]
@mattreaganmozilla, the same steps: Opening and closing the tab tray.

**Comment by mattreaganmozilla** [comment]
> the same steps: Opening and closing the tab tray.

Ok interesting. For reference, I'm profiling on device (iPhone 12 mini) and if open/close the tray 10x these are the results I currently get:

Before
Screenshot 2024-11-08 at 1 16 16 PM

After
Screenshot 2024-11-08 at 1 14 44 PM

So this just an approximation based on the output from Allocations, but across 10x repetitions I see about a ~1.2MB total increase in memory (or ~122kb each open/close) with the fix in place. Without the fix it's closer to ~25-30MB total (or about 2-3MB each open/close). So that's a factor of 20x or more.

In any case, I'd like to get the leak to zero so I'm going to keep this open a bit longer.

**Comment by mattreaganmozilla** [comment]
Note: this is still on my radar, but I haven't had time yet to further investigate into the root culprit. I may merge this temporary fix for v134 since it should significantly reduce the impact of the memory leak for users. I'm planning to keep the ticket open regardless so that we can eventually replace this fix with a better suitable long-term solution.

**Comment by PARAIPAN9** [comment]
> Note: this is still on my radar, but I haven't had time yet to further investigate into the root culprit. I may merge this temporary fix for v134 since it should significantly reduce the impact of the memory leak for users. I'm planning to keep the ticket open regardless so that we can eventually replace this fix with a better suitable long-term solution.

Sure, thanks for the update @mattreaganmozilla!!

**Pull request merged**